File: with_haskell/lib/Day3.hs
module Day3 (part1, part2) where

import Text.Regex.TDFA ((=~))
import Data.Maybe (fromMaybe)

part1 :: [String] -> Int
part1 = sum . extractMuls . concat
  where
    extractMuls :: String -> [Int]
    extractMuls input =
        [ read a * read b
        | [_, a, b] <- input =~ "mul\\(([0-9]{1,3}),([0-9]{1,3})\\)" :: [[String]]
        ]

part2 :: [String] -> Int
part2 = fst . foldl processMatch (0, True) . findMatches . concat
  where
    findMatches :: String -> [[String]]
    findMatches t = t =~ "(do|don't)\\(\\)|mul\\(([0-9]{1,3}),([0-9]{1,3})\\)" :: [[String]]

    processMatch :: (Int, Bool) -> [String] -> (Int, Bool)
    processMatch (sum, _) [_, "do", "", ""] = (sum, True)
    processMatch (sum, _) [_, "don't", "", ""] = (sum, False)
    processMatch (sum, True) [_, "", a, b] = (sum + read a * read b, True)
    processMatch state _ = state

File: with_elixir/lib/day3.ex
defmodule Day3 do
  def part1(content) do
    content
    |> Enum.join()
    |> process_mul()
  end

  def part2(content) do
    content
    |> Enum.join()
    |> process_conditional_mul()
  end

  defp process_mul(input) do
    Regex.scan(~r/mul\((\d{1,3}),(\d{1,3})\)/, input, capture: :all_but_first)
    |> Enum.map(fn [a, b] -> String.to_integer(a) * String.to_integer(b) end)
    |> Enum.sum()
  end

  defp process_conditional_mul(input) do
    Regex.scan(~r/(do|don't)\(\)|mul\((\d{1,3}),(\d{1,3})\)/, input, capture: :all_but_first)
    |> Enum.reduce({0, true}, fn
      ["do"], {sum, _} -> {sum, true}
      ["don't"], {sum, _} -> {sum, false}
      ["", a, b], {sum, true} -> {sum + String.to_integer(a) * String.to_integer(b), true}
      _, state -> state
    end)
    |> elem(0)
  end
end
